data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/plugins/InMemorySource.java

```diff
@@ -57,20 +78,55 @@ private class SourceRunner implements Runnable {
             this.buffer = buffer;
         }
 
+        protected void writeToBuffer(final List<Record<Event>> records) throws Exception {
+            if(!records.isEmpty()) {
+                buffer.writeAll(records, 200);
+            } else {
+                Thread.sleep(10);
+            }
+        }
+
         @Override
         public void run() {
             while (!isStopped) {
                 try {
                     final List<Record<Event>> records = inMemorySourceAccessor.read(testingKey);
-                    if(!records.isEmpty()) {
-                        buffer.writeAll(records, 200);
-                    } else {
-                        Thread.sleep(10);
-                    }
+                    writeToBuffer(records);
                 } catch (final Exception ex) {
                     LOG.error("Error during source loop.", ex);
                 }
             }
         }
     }
+
+    private class AcknowledgementSourceRunner extends SourceRunner {
+        private final AcknowledgementSetManager acknowledgementSetManager;
+        private AtomicBoolean ackReceived;
+
+        AcknowledgementSourceRunner(final AcknowledgementSetManager acknowledgementSetManager, final Buffer<Record<Event>> buffer) {
+            super(buffer);
+            this.acknowledgementSetManager = acknowledgementSetManager;
+            this.ackReceived = new AtomicBoolean(false);
+        }
+
+        @Override
+        public void run() {
+            while (!isStopped) {
+                try {
+                    final List<Record<Event>> records = inMemorySourceAccessor.read(testingKey);
+                    AcknowledgementSet ackSet =
+                            acknowledgementSetManager.create((result) ->
+                                {
+                                    inMemorySourceAccessor.setAckReceived(result);
+                                },
+                                Duration.ofSeconds(5));
+                    records.stream().forEach((record) -> { ackSet.add(record.getData()); });
+                    writeToBuffer(records);
+                } catch (final Exception ex) {
```

Comment:
The `InMemorySource` is just used for testing? If so should we just throw this exception so that it is more obvious on a failed test?